File: Runtime/CoreLib/UInt64.cs
// UInt64.cs
// Script#/Libraries/CoreLib
// This source code is subject to terms and conditions of the Apache License, Version 2.0.
//

using System;
using System.Runtime.CompilerServices;

namespace System
{
    /// <summary>
    /// The ulong data type which is mapped to the Number type in Javascript.
    /// </summary>
    [ScriptNamespace("ss")]
    [ScriptName("UInt64")]
    [Imported(ObeysTypeSystem = true)]
    public struct UInt64 : IComparable<UInt64>, IEquatable<UInt64>, IFormattable
    {
        internal readonly int Low;
        internal readonly int Mid;
        internal readonly int High;

        [InlineCode("{$System.UInt64}.Zero")]
        private UInt64(DummyTypeUsedToAddAttributeToDefaultValueTypeConstructor _)
        {
            Low = 0;
            Mid = 0;
            High = 0;
        }

        private UInt64(int low, int mid, int high)
        {
            Low = low & 0xffffff;
            Mid = mid & 0xffffff;
            High = high & 0xffff;
        }

        public static readonly UInt64 MinValue = new UInt64(0, 0, 0);

        public static readonly UInt64 Zero = new UInt64(0, 0, 0);

        public static readonly UInt64 One = new UInt64(1, 0, 0);

        public static readonly UInt64 MaxValue = new UInt64(0xffffff, 0xffffff, 0x7fff);

        public string Format(string format)
        {
            return ToString(format);
        }

        public string LocaleFormat(string format)
        {
            return ToString(format);
        }

        public static UInt64 Parse(string text)
        {
            UInt64 result;
            if (!TryParse(text, out result))
                throw new FormatException("Input string was not in a correct format.");

            return result;
        }

        public static bool TryParse(string text, out UInt64 result)
        {
            const int radix = 10;
            result = Zero;

            //if (style & System.Globalization.NumberStyles.AllowHexSpecifier)
            //    radix = 16;

            UInt64 rdx = new UInt64(radix, 0, 0);

            for (var i = 0; i < text.Length; i++)
            {
                if (i == 0 && text[i] == '-')
                {
                    result = Zero;
                    return false;
                }
                Int32 c = Int32.Parse((string)text[i], radix);
                if (Double.IsNaN(c))
                {
                    result = Zero;
                    return false;
                }
                result = new UInt64(c, 0, 0) + (rdx * result);
            }

            return true;
        }

        public override string ToString()
        {
            return ToString(null);
        }

        public string ToString(string format)
        {
            UInt64 ten = new UInt64(10, 0, 0);
            UInt64 a = this;

            string s = "";

            do
            {
                UInt64 r = a % ten;
                s = r.Low.ToString() + s;
                a = a / ten;
            } while (a > Zero);

            return s;
        }

        public int CompareTo(UInt64 other)
        {
            if (this < other) return -1;
            return this > other ? 1 : 0;
        }

        public bool Equals(UInt64 other)
        {
            return Low == other.Low && Mid == other.Mid && High == other.High;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is UInt64 && Equals((UInt64)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hashCode = Low;
                hashCode = (hashCode * 397) ^ Mid;
                hashCode = (hashCode * 397) ^ High;
                return hashCode;
            }
        }

        private const int Mask = unchecked((int)0xffffff000000);
        public static UInt64 operator +(UInt64 a, UInt64 b)
        {
            //same as Int64
            int cLow = a.Low + b.Low;
            int rLow = (cLow & Mask) >> 24;
            int cMid = rLow + a.Mid + b.Mid;
            int rMid = (cMid & Mask) >> 24;
            int cHigh = rMid + a.High + b.High;

            return new UInt64(cLow, cMid, cHigh);
        }

        public static UInt64 operator -(UInt64 a, UInt64 b)
        {
            //same as Int64
            int cLow = (a.Low - b.Low) | 0;
            int rLow = 0;
            if (cLow < 0)
            {
                cLow = 0x1000000 + cLow;
                rLow = -1;
            }
            int cMid = (rLow + ((a.Mid - b.Mid) | 0)) | 0;
            int rMid = 0;
            if (cMid < 0)
            {
                cMid = 0x1000000 + cMid;
                rMid = -1;
            }
            int cHigh = (rMid + ((a.High - b.High) | 0)) | 0;
            if (cHigh < 0)
            {
                cHigh = 0x10000 + cHigh;
            }

            return new UInt64(cLow, cMid, cHigh);
        }

        public static UInt64 operator *(UInt64 a, UInt64 b)
        {
            if (a.Equals(Zero) || b.Equals(Zero))
                return Zero;

            if (a > b)
                return b * a;

            UInt64 c = Zero;

            if ((a.Low & 1) == 1)
            {
                c = b;
            }

            while (a != One)
            {
                a >>= 1;
                b <<= 1;

                if ((a.Low & 1) == 1)
                    c += b;
            }

            return c;
        }

        public static UInt64 operator /(UInt64 a, UInt64 b)
        {

            if (b.Equals(MinValue))
                throw new DivideByZeroException();

            UInt64 q = Zero;
            UInt64 r = Zero;

            for (int i = 63; i >= 0; i--)
            {
                r <<= 1;
                int x;
                int s;
                if (i < 24)
                {
                    x = a.Low;
                    s = i;
                }
                else if (i < 48)
                {
                    x = a.Mid;
                    s = i - 24;
                }
                else
                {
                    x = a.High;
                    s = i - 48;
                }

                r = new UInt64(r.Low | ((x & (1 << s)) >> s), r.Mid, r.High);

                if (r < b) continue;

                r -= b;

                if (i < 24)
                    q = new UInt64(q.Low | (1 << s), q.Mid, q.High);
                else if (i < 48)
                    q = new UInt64(q.Low, q.Mid | (1 << s), q.High);
                else
                    q = new UInt64(q.Low, q.Mid, q.High | (1 << s));
            }

            return q;
        }

        public static UInt64 operator %(UInt64 a, UInt64 b)
        {
            if (b.Equals(MinValue))
                throw new DivideByZeroException();

            UInt64 r = Zero;

            for (int i = 63; i >= 0; i--)
            {
                r <<= 1;
                int x;
                int s;
                if (i < 24)
                {
                    x = a.Low;
                    s = i;
                }
                else if (i < 48)
                {
                    x = a.Mid;
                    s = i - 24;
                }
                else
                {
                    x = a.High;
                    s = i - 48;
                }

                r = new UInt64(r.Low | ((x & (1 << s)) >> s), r.Mid, r.High);

                if (r >= b)
                    r -= b;
            }

            return r;
        }

        public static UInt64 operator &(UInt64 a, UInt64 b)
        {
            //same as Int64
            return new UInt64(a.Low & b.Low, a.Mid & b.Mid, a.High & b.High);
        }

        public static UInt64 operator |(UInt64 a, UInt64 b)
        {
            //same as Int64
            return new UInt64(a.Low | b.Low, a.Mid | b.Mid, a.High | b.High);
        }

        public static UInt64 operator ^(UInt64 a, UInt64 b)
        {
            //same as Int64
            return new UInt64(a.Low ^ b.Low, a.Mid ^ b.Mid, a.High ^ b.High);
        }

        public static UInt64 operator <<(UInt64 a, int b)
        {
            //same as Int64
            b = b & 0x3f;

            const int maxShift = 8;

            if (b > 8)
                return (a << maxShift) << (b - maxShift);

            int cLowT = a.Low << b;
            int cLow = cLowT & 0xffffff;
            int rLow = (int)((uint)cLowT >> 24) & 0xffffff;
            int cMidT = (a.Mid << b) | rLow;
            int cMid = cMidT & 0xffffff;
            int rMid = (int)((uint)cMidT >> 24) & 0xffff;
            int cHighT = a.High << b;
            int cHigh = (cHighT & 0xffff) | rMid;

            return new UInt64(cLow, cMid, cHigh);
        }

        public static UInt64 operator >>(UInt64 a, int b)
        {
            b = b & 0x3f;

            if (b > 24)
                return (a >> 24) >> (b - 24);

            int m = (1 << b) - 1;
            int rHigh = (a.High & m) << (24 - b);
            int cHighT = a.High >> b;
            int rMid = (a.Mid & m) << (24 - b);
            int cMidT = a.Mid >> b;
            int cLowT = a.Low >> b;

            return new UInt64(cLowT | rMid, cMidT | rHigh, cHighT);
        }

        public static bool operator ==(UInt64 a, UInt64 b)
        {
            //same as Int64
            return a.Low == b.Low && a.Mid == b.Mid && a.High == b.High;
        }

        public static bool operator !=(UInt64 a, UInt64 b)
        {
            //same as Int64
            return a.Low != b.Low || a.Mid != b.Mid || a.High != b.High;
        }

        public static bool operator <=(UInt64 a, UInt64 b)
        {
            int adiff = a.High - b.High;
            if (adiff < 0)
                return true;

            if (adiff > 0)
                return false;

            int bdiff = a.Mid - b.Mid;
            if (bdiff < 0)
                return true;

            if (bdiff > 0)
                return false;

            return a.Low <= b.Low;
        }

        public static bool operator >=(UInt64 a, UInt64 b)
        {
            int adiff = a.High - b.High;
            if (adiff > 0)
                return true;

            if (adiff < 0)
                return false;

            int bdiff = a.Mid - b.Mid;
            if (bdiff > 0)
                return true;

            if (bdiff < 0)
                return false;

            return a.Low >= b.Low;
        }

        public static bool operator <(UInt64 a, UInt64 b)
        {
            int adiff = a.High - b.High;
            if (adiff < 0)
                return true;

            if (adiff > 0)
                return false;

            int bdiff = a.Mid - b.Mid;
            if (bdiff < 0)
                return true;

            if (bdiff > 0)
                return false;

            return a.Low < b.Low;
        }

        public static bool operator >(UInt64 a, UInt64 b)
        {
            int adiff = a.High - b.High;
            if (adiff > 0)
                return true;

            if (adiff < 0)
                return false;

            int bdiff = a.Mid - b.Mid;
            if (bdiff > 0)
                return true;

            if (bdiff < 0)
                return false;

            return a.Low > b.Low;
        }

        [InlineCode("{a}")]
        public static UInt64 operator +(UInt64 a)
        {
            return a;
        }

        public static Int64 operator -(UInt64 a)
        {
            return (Int64)~a + Int64.One;
        }

        public static UInt64 operator ~(UInt64 a)
        {
            return new UInt64(~a.Low, ~a.Mid, ~a.High);
        }

        public static UInt64 operator ++(UInt64 a)
        {
            //same as Int64
            int cLow = a.Low + 1;
            int rLow = (cLow & Mask) >> 24;
            int cMid = rLow + a.Mid;
            int rMid = (cMid & Mask) >> 24;
            int cHigh = rMid + a.High;

            return new UInt64(cLow, cMid, cHigh);
        }

        public static UInt64 operator --(UInt64 a)
        {
            //same as Int64
            int cLow = (a.Low - 1) | 0;
            int rLow = 0;
            if (cLow < 0)
            {
                cLow = 0x1000000 + cLow;
                rLow = -1;
            }
            int cMid = (rLow + a.Mid) | 0;
            int rMid = 0;
            if (cMid < 0)
            {
                cMid = 0x1000000 + cMid;
                rMid = -1;
            }
            int cHigh = (rMid + a.High) | 0;
            if (cHigh < 0)
            {
                cHigh = 0x10000 + cHigh;
            }

            return new UInt64(cLow, cMid, cHigh);
        }

        //TODO Casts
        public static explicit operator UInt64(Int64 a)
        {
            return new UInt64(a.Low, a.Mid, a.High);
        }

        public static implicit operator UInt64(Byte a)
        {
            return new UInt64(a, 0, 0);
        }

        public static implicit operator UInt64(SByte a)
        {
            return new UInt64(a, a < 0 ? 0xffffff : 0, a < 0 ? 0xffff : 0);
        }

        public static implicit operator UInt64(UInt16 a)
        {
            return new UInt64(a, 0, 0);
        }

        public static implicit operator UInt64(Int16 a)
        {
            return new UInt64(a, a < 0 ? 0xffffff : 0, a < 0 ? 0xffff : 0);
        }

        public static implicit operator UInt64(UInt32 a)
        {
            return new UInt64((Int32)a, (Int32)(a >> 24), 0);
        }

        public static implicit operator UInt64(Int32 a)
        {
            return new UInt64(a, a >> 24, a < 0 ? 0xffff : 0);
        }

        public static explicit operator UInt64(Double a)
        {
            if (a < 0)
                throw new ArgumentOutOfRangeException();

            double floorN = Math.Floor(a);
            int n0 = (int)floorN | 0;
            int n1 = (int)(floorN / 0x1000000) | 0;
            int n2 = (int)(floorN / 281474976710656.0) | 0;

            return new UInt64(n0, n1, n2);
        }

        public static explicit operator UInt64(Single a)
        {
            return (UInt64)(double)a;
        }

        public static explicit operator UInt64(Decimal a)
        {
            return (UInt64)(double)a;
        }

        public static implicit operator Byte(UInt64 a)
        {
            return (Byte)(a.Low & Byte.MaxValue);
        }

        public static implicit operator SByte(UInt64 a)
        {
            return (SByte)(a.Low & Byte.MaxValue);
        }

        public static implicit operator UInt16(UInt64 a)
        {
            return (UInt16)(a.Low & UInt16.MaxValue);
        }

        public static implicit operator Int16(UInt64 a)
        {
            return (Int16)(a.Low & UInt16.MaxValue);
        }

        public static implicit operator UInt32(UInt64 a)
        {
            //return (UInt32)((a.Low | a.Mid << 24) & UInt32.MaxValue);
            // return (a.$low | a.$mid << 24) & 4294967295;
            throw new NotImplementedException();
        }

        public static implicit operator Int32(UInt64 a)
        {
            //return (UInt32)((a.Low | a.Mid << 24) & UInt32.MaxValue);
            // return (a.$low | a.$mid << 24) & 4294967295;
            throw new NotImplementedException();
        }

        public static explicit operator Double(UInt64 a)
        {
            return 16777216.0 * ((16777216.0 * a.High) + a.Mid) + a.Low;
        }

        public static explicit operator Single(UInt64 a)
        {
            return 16777216f * ((16777216f * a.High) + a.Mid) + a.Low;
        }

        public static explicit operator Decimal(UInt64 a)
        {
            return 16777216m * ((16777216m * a.High) + a.Mid) + a.Low;
        }
    }
}

File: Runtime/CoreLib.Script/UInt64.js
////////////////////////////////////////////////////////////////////////////////
// ss.UInt64
var ss_UInt64 = function (low, mid, high) {
    this.$low = 0;
    this.$mid = 0;
    this.$high = 0;
    this.$low = low & 16777215;
    this.$mid = mid & 16777215;
    this.$high = high & 65535;
};
ss_UInt64.__typeName = 'ss.UInt64';
ss_UInt64.createInstance = function () {
    return ss.Int32.Zero;
};
ss_UInt64.getDefaultValue = ss_UInt64.createInstance = function () {
	return ss_UInt64.zero;
};;
ss_UInt64.parse = function (text, radix) {
	var result = {};
	if (!ss_UInt64.tryParse(text, radix, result)) {
		throw new ss.FormatException('Input string was not in a correct format.');
	}
	return result.$;
};
ss_UInt64.tryParse = function (text, radix, result) {
	if (radix < 2 || radix > 36) {
		throw new ss.ArgumentOutOfRangeException('radix', 'radix argument must be between 2 and 36');
	}
	result.$ = ss_UInt64.zero;
	//if (style & System.Globalization.NumberStyles.AllowHexSpecifier)
	//    radix = 16;
	var rdx = new ss_UInt64(radix, 0, 0);
	for (var i = 0; i < text.length; i++) {
		if (i === 0 && text.charCodeAt(i) === 45) {
			result.$ = ss_UInt64.zero;
			return false;
		}
		var c = parseInt(String.fromCharCode(text.charCodeAt(i)), radix);
		if (isNaN(c)) {
			result.$ = ss_UInt64.zero;
			return false;
		}
		result.$ = ss_UInt64.op_Addition(new ss_UInt64(c, 0, 0), ss_UInt64.op_Multiply(rdx, result.$));
	}
	return true;
};
ss_UInt64.op_Addition = function (a, b) {
	//same as Int64
	var cLow = a.$low + b.$low;
	var rLow = (cLow & ss_UInt64.$mask) >> 24;
	var cMid = rLow + a.$mid + b.$mid;
	var rMid = (cMid & ss_UInt64.$mask) >> 24;
	var cHigh = rMid + a.$high + b.$high;
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.op_Subtraction = function (a, b) {
	//same as Int64
	var cLow = a.$low - b.$low | 0;
	var rLow = 0;
	if (cLow < 0) {
		cLow = 16777216 + cLow;
		rLow = -1;
	}
	var cMid = rLow + (a.$mid - b.$mid | 0) | 0;
	var rMid = 0;
	if (cMid < 0) {
		cMid = 16777216 + cMid;
		rMid = -1;
	}
	var cHigh = rMid + (a.$high - b.$high | 0) | 0;
	if (cHigh < 0) {
		cHigh = 65536 + cHigh;
	}
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.op_Multiply = function (a, b) {
	if (a.equalsT(ss_UInt64.zero) || b.equalsT(ss_UInt64.zero)) {
		return ss_UInt64.zero;
	}
	if (ss_UInt64.op_GreaterThan(a, b)) {
		return ss_UInt64.op_Multiply(b, a);
	}
	var c = ss_UInt64.zero;
	if ((a.$low & 1) === 1) {
		c = b;
	}
	while (ss_UInt64.op_Inequality(a, ss_UInt64.one)) {
		a = ss_UInt64.op_RightShift(a, 1);
		b = ss_UInt64.op_LeftShift(b, 1);
		if ((a.$low & 1) === 1) {
			c = ss_UInt64.op_Addition(c, b);
		}
	}
	return c;
};
ss_UInt64.op_Division = function (a, b) {
	if (b.equalsT(ss_UInt64.minValue)) {
		throw new ss.DivideByZeroException();
	}
	var q = ss_UInt64.zero;
	var r = ss_UInt64.zero;
	for (var i = 63; i >= 0; i--) {
		r = ss_UInt64.op_LeftShift(r, 1);
		var x;
		var s;
		if (i < 24) {
			x = a.$low;
			s = i;
		}
		else if (i < 48) {
			x = a.$mid;
			s = i - 24;
		}
		else {
			x = a.$high;
			s = i - 48;
		}
		r = new ss_UInt64(r.$low | (x & 1 << s) >> s, r.$mid, r.$high);
		if (ss_UInt64.op_LessThan(r, b)) {
			continue;
		}
		r = ss_UInt64.op_Subtraction(r, b);
		if (i < 24) {
			q = new ss_UInt64(q.$low | 1 << s, q.$mid, q.$high);
		}
		else if (i < 48) {
			q = new ss_UInt64(q.$low, q.$mid | 1 << s, q.$high);
		}
		else {
			q = new ss_UInt64(q.$low, q.$mid, q.$high | 1 << s);
		}
	}
	return q;
};
ss_UInt64.op_Modulus = function (a, b) {
	if (b.equalsT(ss_UInt64.minValue)) {
		throw new ss.DivideByZeroException();
	}
	var r = ss_UInt64.zero;
	for (var i = 63; i >= 0; i--) {
		r = ss_UInt64.op_LeftShift(r, 1);
		var x;
		var s;
		if (i < 24) {
			x = a.$low;
			s = i;
		}
		else if (i < 48) {
			x = a.$mid;
			s = i - 24;
		}
		else {
			x = a.$high;
			s = i - 48;
		}
		r = new ss_UInt64(r.$low | (x & 1 << s) >> s, r.$mid, r.$high);
		if (ss_UInt64.op_GreaterThanOrEqual(r, b)) {
			r = ss_UInt64.op_Subtraction(r, b);
		}
	}
	return r;
};
ss_UInt64.op_BitwiseAnd = function (a, b) {
	//same as Int64
	return new ss_UInt64(a.$low & b.$low, a.$mid & b.$mid, a.$high & b.$high);
};
ss_UInt64.op_BitwiseOr = function (a, b) {
	//same as Int64
	return new ss_UInt64(a.$low | b.$low, a.$mid | b.$mid, a.$high | b.$high);
};
ss_UInt64.op_ExclusiveOr = function (a, b) {
	//same as Int64
	return new ss_UInt64(a.$low ^ b.$low, a.$mid ^ b.$mid, a.$high ^ b.$high);
};
ss_UInt64.op_LeftShift = function (a, b) {
	b = b & 63;
	if (b === 0) {
		return a;
	}
	var cLow, cMid, cHigh;
	if (b <= 24) {
		cLow = a.$low << b;
		cMid = a.$low >> 24 - b | a.$mid << b;
		cHigh = a.$mid >> 24 - b | a.$high << b;
	}
	else if (b <= 48) {
		cLow = 0;
		cMid = a.$low << b - 24;
		cHigh = a.$low >> 48 - b | a.$mid << b - 24;
	}
	else {
		cLow = 0;
		cMid = 0;
		cHigh = a.$low << b - 48;
	}
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.op_RightShift = function (a, b) {
	b = b & 63;
	if (b === 0) {
		return a;
	}
	var cLow, cMid, cHigh;
	if (b <= 24) {
		cLow = a.$mid << 24 - b | a.$low >> b;
		cMid = a.$high << 24 - b | a.$mid >> b;
		cHigh = a.$high >> b;
	}
	else if (b <= 48) {
		cLow = a.$high << 48 - b | a.$mid >> b - 24;
		cMid = a.$high >> b - 24;
		cHigh = 0;
	}
	else {
		cLow = a.$high >> b - 48;
		cMid = 0;
		cHigh = 0;
	}
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.op_Equality = function (a, b) {
	//same as Int64
	return a.$low === b.$low && a.$mid === b.$mid && a.$high === b.$high;
};
ss_UInt64.op_Inequality = function (a, b) {
	//same as Int64
	return a.$low !== b.$low || a.$mid !== b.$mid || a.$high !== b.$high;
};
ss_UInt64.op_LessThanOrEqual = function (a, b) {
	var adiff = a.$high - b.$high;
	if (adiff < 0) {
		return true;
	}
	if (adiff > 0) {
		return false;
	}
	var bdiff = a.$mid - b.$mid;
	if (bdiff < 0) {
		return true;
	}
	if (bdiff > 0) {
		return false;
	}
	return a.$low <= b.$low;
};
ss_UInt64.op_GreaterThanOrEqual = function (a, b) {
	var adiff = a.$high - b.$high;
	if (adiff > 0) {
		return true;
	}
	if (adiff < 0) {
		return false;
	}
	var bdiff = a.$mid - b.$mid;
	if (bdiff > 0) {
		return true;
	}
	if (bdiff < 0) {
		return false;
	}
	return a.$low >= b.$low;
};
ss_UInt64.op_LessThan = function (a, b) {
	var adiff = a.$high - b.$high;
	if (adiff < 0) {
		return true;
	}
	if (adiff > 0) {
		return false;
	}
	var bdiff = a.$mid - b.$mid;
	if (bdiff < 0) {
		return true;
	}
	if (bdiff > 0) {
		return false;
	}
	return a.$low < b.$low;
};
ss_UInt64.op_GreaterThan = function (a, b) {
	var adiff = a.$high - b.$high;
	if (adiff > 0) {
		return true;
	}
	if (adiff < 0) {
		return false;
	}
	var bdiff = a.$mid - b.$mid;
	if (bdiff > 0) {
		return true;
	}
	if (bdiff < 0) {
		return false;
	}
	return a.$low > b.$low;
};
ss_UInt64.op_UnaryNegation = function (a) {
	return ss_Int64.op_Addition(ss_Int64.fromUInt64(ss_UInt64.op_OnesComplement(a)), ss_Int64.one);
};
ss_UInt64.op_OnesComplement = function (a) {
	return new ss_UInt64(~a.$low, ~a.$mid, ~a.$high);
};
ss_UInt64.op_Increment = function (a) {
	//same as Int64
	var cLow = a.$low + 1;
	var rLow = (cLow & ss_UInt64.$mask) >> 24;
	var cMid = rLow + a.$mid;
	var rMid = (cMid & ss_UInt64.$mask) >> 24;
	var cHigh = rMid + a.$high;
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.op_Decrement = function (a) {
	//same as Int64
	var cLow = a.$low - 1 | 0;
	var rLow = 0;
	if (cLow < 0) {
		cLow = 16777216 + cLow;
		rLow = -1;
	}
	var cMid = rLow + a.$mid | 0;
	var rMid = 0;
	if (cMid < 0) {
		cMid = 16777216 + cMid;
		rMid = -1;
	}
	var cHigh = rMid + a.$high | 0;
	if (cHigh < 0) {
		cHigh = 65536 + cHigh;
	}
	return new ss_UInt64(cLow, cMid, cHigh);
};
ss_UInt64.fromInt64 = function (a) {
	return new ss_UInt64(a.$low, a.$mid, a.$high);
};
ss_UInt64.fromUInt32 = function (a) {
	return new ss_UInt64(a, a >>> 24, 0);
};
ss_UInt64.fromInt32 = function (a) {
	return new ss_UInt64(a, a >> 24, ((a < 0) ? 65535 : 0));
};
ss_UInt64.fromDouble = function (a) {
	if (a < 0) {
		throw new ss.ArgumentOutOfRangeException();
	}
	var floorN = Math.floor(a);
	var n0 = ss.Int32.trunc(floorN) | 0;
	var n1 = ss.Int32.trunc(floorN / 16777216) | 0;
	var n2 = ss.Int32.trunc(floorN / 281474976710656) | 0;
	return new ss_UInt64(n0, n1, n2);
};
ss_UInt64.fromSingle = function (a) {
	return ss_UInt64.fromDouble(a);
};
ss_UInt64.fromDecimal = function (a) {
	return ss_UInt64.fromDouble(a);
};
ss_UInt64.toUInt32 = function (a) {
	return (a.$low | a.$mid << 24) & 4294967295;
};
ss_UInt64.toInt32 = function (a) {
	return (a.$low | a.$mid << 24) & 4294967295;
};
ss_UInt64.toDouble = function (a) {
	return 16777216 * (16777216 * a.$high + a.$mid) + a.$low;
};
ss_UInt64.toSingle = function (a) {
	return 16777216 * (16777216 * a.$high + a.$mid) + a.$low;
};
ss_UInt64.toDecimal = function (a) {
	return 16777216 * (16777216 * a.$high + a.$mid) + a.$low;
};
ss.initClass(ss_UInt64, ss, {
	format$1: function (format) {
		return this.format(format);
	},
	localeFormat: function (format) {
		return this.format(format);
	},
	toString: function () {
		return this.format(null);
	},
	format: function (format) {
		var ten = new ss_UInt64(10, 0, 0);
		var a = this;
		var s = '';
		do {
			var r = ss_UInt64.op_Modulus(a, ten);
			s = r.$low.toString() + s;
			a = ss_UInt64.op_Division(a, ten);
		} while (ss_UInt64.op_GreaterThan(a, ss_UInt64.zero));
		return s;
	},
	toString$1: function (radix) {
		if (radix < 2 || radix > 36) {
			throw new ss.ArgumentOutOfRangeException('radix', 'radix argument must be between 2 and 36');
		}
		var rad = new ss_UInt64(radix, 0, 0);
		var a = this;
		var s = '';
		do {
			var r = ss_UInt64.op_Modulus(a, rad);
			s = r.$low.toString(radix) + s;
			a = ss_UInt64.op_Division(a, rad);
		} while (ss_UInt64.op_GreaterThan(a, ss_UInt64.zero));
		return s;
	},
	compareTo: function (other) {
		if (ss_UInt64.op_LessThan(this, other)) {
			return -1;
		}
		return (ss_UInt64.op_GreaterThan(this, other) ? 1 : 0);
	},
	equalsT: function (other) {
		return this.$low === other.$low && this.$mid === other.$mid && this.$high === other.$high;
	},
	equals: function (obj) {
		if (ss.referenceEquals(null, obj)) {
			return false;
		}
		return ss.isInstanceOfType(obj, ss_UInt64) && this.equalsT(ss.unbox(ss.cast(obj, ss_UInt64)));
	},
	getHashCode: function () {
		var hashCode = this.$low;
		hashCode = hashCode * 397 ^ this.$mid;
		hashCode = hashCode * 397 ^ this.$high;
		return hashCode;
	}
}, Object, [ss.IComparable, ss.IEquatable, ss.IFormattable]);
ss_UInt64.__class = false;
(function () {
    ss_UInt64.minValue = new ss_UInt64(0, 0, 0);
    ss_UInt64.zero = new ss_UInt64(0, 0, 0);
    ss_UInt64.one = new ss_UInt64(1, 0, 0);
    ss_UInt64.maxValue = new ss_UInt64(16777215, 16777215, 32767);
    ss_UInt64.$mask = -16777216;
})();